### PR TITLE
entproto/cmd/protoc-gen-entgrpc: starting to migrate code-generation to be template based

### PR DESCRIPTION
**Pull request by rotemtam**
The changes made in #96 made me realize that the mixed imperative/template-based approach to code-gen in this project isn't working well enough. Code has become convoluted and hard to read (for me at least). 

This PR starts a change that will be spread across a few extra PRs to use an approach closer to the rest of the `ent` code-gen world - using `text/template`. A simple `ident` function is pushed into the template context to allow template authors to register go ident's so they are automatically qualified and imported by `protogen`'s facilities. 

The change is made gradually so the with the current PR both styles are used. 

cc @tarrencev

### CODE REVIEW

File: entproto/cmd/protoc-gen-entgrpc/generate.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+//go:generate go run github.com/go-bindata/go-bindata/go-bindata -o=internal/bindata.go -pkg=internal -modtime=1 ./template
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/generate.go:17]: Move the go generate close to its usage (`var templates`).

File: entproto/cmd/protoc-gen-entgrpc/main.go [hunk 2]
```diff
@@ -90,100 +92,56 @@ func newServiceGenerator(plugin *protogen.Plugin, file *protogen.File, graph *ge
...
-	g.P()
-	g.generateConstructor()
-	g.P()
-	if err := g.generateEnumConvertFuncs(); err != nil {
+	tmpl := template.New("").
+		Funcs(gen.Funcs).
+		Funcs(template.FuncMap{
+			"ident": func(ident protogen.GoIdent) string {
+				return g.QualifiedGoIdent(ident)
+			},
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/main.go:109]:
```suggestion
			"ident": g.QualifiedGoIdent,
```

File: entproto/cmd/protoc-gen-entgrpc/main.go [hunk 3]
```diff
@@ -297,6 +255,13 @@ func extractEntTypeName(s *protogen.Service, g *gen.Graph) (*gen.Type, error) {
...
+	ip := path.Join(string(g.EntPackage), subpath)
 	return protogen.GoImportPath(ip).Ident(ident)
 }
+
+var (
+	templates = [][]byte{
+		internal.MustAsset("template/service.tmpl"),
+		internal.MustAsset("template/enums.tmpl"),
+	}
+)
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/main.go:267]:
```suggestion
var templates = [][]byte{
	internal.MustAsset("template/service.tmpl"),
	internal.MustAsset("template/enums.tmpl"),
}
```

File: entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{{ define "enums" }}
+    {{- $root := . -}}
+    {{ range .FieldMap.Enums }}
+        {{- $enumType := .PbFieldDescriptor.GetEnumType -}}
+        {{- $enumName := print $root.EntType.Name "_" $enumType.GetName -}}
+        {{- $pbEnumIdent := $root.File.GoImportPath.Ident $enumName   -}}
+        {{- $entLcase := snake $root.EntType.Name -}}
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:7]:
```suggestion
    {{ range .FieldMap.Enums }}
        {{- $enumType := .PbFieldDescriptor.GetEnumType -}}
        {{- $enumName := print $.EntType.Name "_" $enumType.GetName -}}
        {{- $pbEnumIdent := $.File.GoImportPath.Ident $enumName   -}}
        {{- $entLcase := snake $.EntType.Name -}}
```

Comment (a8m) [reply entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:7]: BTW, you don't need all these `-` in the top-level declaration if you're running `go/format`/`imports` at the end.

File: entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        {{- $entEnumIdent := entIdent $entLcase .PbStructField -}}
+        func toProto{{ $pbEnumIdent.GoName }} (e {{ $entEnumIdent | ident }}) {{ $pbEnumIdent | ident }} {
+            if v, ok := {{ $pbEnumIdent.GoName }}_value[{{ qualify "strings" "ToUpper"}}(string(e))]; ok {
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:10]:
```suggestion
            if v, ok := {{ $pbEnumIdent.GoName }}_value[{{ qualify "strings" "ToUpper" }}(string(e))]; ok {
```

File: entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl [hunk 6]
```diff
@@ -0,0 +1,23 @@
+{{ define "enums" }}
+    {{- $root := . -}}
+    {{ range .FieldMap.Enums }}
+        {{- $enumType := .PbFieldDescriptor.GetEnumType -}}
+        {{- $enumName := print $root.EntType.Name "_" $enumType.GetName -}}
+        {{- $pbEnumIdent := $root.File.GoImportPath.Ident $enumName   -}}
+        {{- $entLcase := snake $root.EntType.Name -}}
+        {{- $entEnumIdent := entIdent $entLcase .PbStructField -}}
+        func toProto{{ $pbEnumIdent.GoName }} (e {{ $entEnumIdent | ident }}) {{ $pbEnumIdent | ident }} {
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:9]:
```suggestion
        func toProto{{ $pbEnumIdent.GoName }} (e {{ ident $entEnumIdent }}) {{ ident $pbEnumIdent }} {
```

Don't need to pipe when you have one func. It also makes the code consistent with other parts.

File: entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+            if v, ok := {{ $pbEnumIdent.GoName }}_value[{{ qualify "strings" "ToUpper"}}(string(e))]; ok {
+                return {{ $pbEnumIdent | ident }}(v)
+            }
+            return {{ $pbEnumIdent | ident }}(0)
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:13]:
```suggestion
                return {{ ident $pbEnumIdent }}(v)
            }
            return {{ ident $pbEnumIdent }}(0)
```

File: entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        }
+
+        func toEnt{{ $pbEnumIdent.GoName }}(e {{$pbEnumIdent | ident}}) {{ $entEnumIdent | ident }} {
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:16]: ditto

File: entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+            if v, ok := {{ $pbEnumIdent.GoName }}_name[int32(e)]; ok {
+                return {{ $entEnumIdent | ident }}({{ qualify "strings" "ToLower" }}(v))
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:18]: ditto

**Review by a8m — APPROVED** [review]
Looks good. Minor comments

File: entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl (hunk 6 above)
Comment (rotemtam) [reply entproto/cmd/protoc-gen-entgrpc/template/enums.tmpl:9]: 🤦

File: entproto/cmd/protoc-gen-entgrpc/main.go (hunk 2 above)
Comment (rotemtam) [reply entproto/cmd/protoc-gen-entgrpc/main.go]: 🤦

**Pull request merged**